### DISCUSSION TITLE
Caching Question

### DISCUSSION THREAD

**Discussion by moisish** [spatie/laravel-permission#1760] [Q&A]
Hello,

From what I can see there is caching on the PermissionRegistrar class which caches all the available permissions and roles on the property $permissions.

What about caching the User's permissions so everytime you use the `hasPermissionTo` method is not running a query against the DB?

Otherwise why does the cache gets cleared on the methods givePermissionTo, revokePermissions etc if the user's permissions are not cached?

**Comment by drbyte** [comment]
Release 4.3.0 updates caching approaches

**Comment by erikn69** [comment]
>What about caching the User's permissions so everytime you use the hasPermissionTo method is not running a query against the DB?

The problem with that is when you has a lot of users, you gonna have a lot of cache keys, they tried something like that once, also it needs a forget especific user cache with unique key, example `spatie.permission.cache.App.Models.User.1` (#926), and now it grows exponentially and everything gets slower(maybe it can be achieved, but it would be necessary to do benchmarks)

>Otherwise why does the cache gets cleared on the methods givePermissionTo, revokePermissions etc if the user's permissions are not cached?

Maybe because `givePermissionTo` and `revokePermissions` are used on `Role.php`, and `Role` is cached with `Permission`
https://github.com/spatie/laravel-permission/blob/78eaa5e06c313a9f3672a7571b4d83b913721b72/src/Models/Role.php#L15-L17
https://github.com/spatie/laravel-permission/blob/78eaa5e06c313a9f3672a7571b4d83b913721b72/src/PermissionRegistrar.php#L121-L123

**Comment by erikn69** [comment]
@drbyte maybe this can be an optional trait for the User's model on the package, i could make a PR, what do you think? 
According laravel debugbar 
```
// Cache
Time for cache: roles (684μs)
Time for cache: permissions (296μs)
// SQL
Time for sql: roles (1.18ms)
Time for sql: permissions (2.03ms)
```

-------
>What about caching the User's permissions so everytime you use the hasPermissionTo method is not running a query against the DB?

@moisish maybe https://stackoverflow.com/a/56263511 can help you for that, only add on `User.php` the next code(or make a trait)

```php
use Illuminate\Database\Eloquent\Collection;
use Spatie\Permission\PermissionRegistrar;
///****///
protected function getPermissionCacheKey(string $relation): string {
    return sprintf(PermissionRegistrar::$cacheKey.'.'.str_replace('\\','.',$this->getMorphClass()).'.%d.'.$relation, $this->getKey());
}
private function getCachedPermissions(string $relation): Collection {
    if ($this->relationLoaded($relation)) {
        return $this->getRelationValue($relation);
    }
    $class = $relation === 'roles' ? app(PermissionRegistrar::class)->getRoleClass() : app(PermissionRegistrar::class)->getPermissionClass();
    $array = \Cache::remember($this->getPermissionCacheKey($relation), config('session.lifetime', 120) * 60, function () use($relation) {
        return $this->getRelationValue($relation)->map(function($data){
            return ['i' => $data->id, 'n' => $data->name, 'g' => $data->guard_name];
        })->all();
    });
    $collection = $class::hydrate(
        collect($array)->map(function ($item) { return ['id' => $item['i'], 'name' => $item['n'], 'guard_name' => $item['g']]; })->all()
    );
    $this->setRelation($relation, $collection);
    return $collection;
}
public function getRolesAttribute(): Collection {
    return $this->getCachedPermissions('roles');
}
public function getPermissionsAttribute(): Collection {
    return $this->getCachedPermissions('permissions');
}
public function forgetAssignedPermissions(): bool {
    return \Cache::forget($this->getPermissionCacheKey('roles')) && \Cache::forget($this->getPermissionCacheKey('permissions'));
}
```
**NOTE:** Don´t forget call `$user->forgetAssignedPermissions();` when assign roles or permissions for refresh cache
